e2e/testcases/otel_collector_test.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/taskgroup"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/pkg/kinds"
+	ocmetrics "kpt.dev/configsync/pkg/metrics"
+)
+
+const (
+	DefaultMonitorKSA = "default"
+	MonitorGSA        = "e2e-test-metric-writer"
+)
+
+func TestOtelCollectorDeployment(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation1, ntopts.RequireGKE(t))
+	nt.T.Cleanup(func() {
+		nt.MustKubectl("delete", "-f", "../testdata/otel-collector/otel-cm-duplicate-timeseries.yaml", "--ignore-not-found")
+		if t.Failed() {
+			nt.PodLogs("config-management-monitoring", ocmetrics.OtelCollectorName, "", false)
+		}
+	})
+
+	// If Workload Identity enabled on cluster, setup KSA to GSA annotation
+	if workloadPool, err := getWorkloadPool(nt); err != nil {
+		nt.T.Fatal(err)
+	} else if workloadPool != "" {
+		nt.T.Log("Workload identity enabled, adding KSA annotation to use %s service account", MonitorGSA)
+		nt.MustKubectl("annotate", "serviceaccount", "--namespace", ocmetrics.MonitoringNamespace, DefaultMonitorKSA,
+			fmt.Sprintf("iam.gke.io/gcp-service-account=%s@%s.iam.gserviceaccount.com", MonitorGSA, *e2e.GCPProject))
+	}
+
+	nt.T.Log("Restart otel-collector pod to refresh the ConfigMap, log and IAM")
+	nomostest.DeletePodByLabel(nt, "app", "opentelemetry", false)
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := validateOtelCollectorStatusCurrent(nt, false); err != nil {
+		nt.T.Fatal()
+	}
+
+	nt.T.Log("Wait 1 minute for metric exporter to settle")
+	time.Sleep(1 * time.Minute)
```

Comment:
What does settle mean here?

Reply:
That when the exporters are either shipping the metrics out, or emitting errors in the log. From the past test there is about 30 to 40 seconds of delay between `otel-collector ready` and `metrics showing up in GCM / errors showing up in pod log`.

Reply:
Is this because of the otel-agent buffer?
I have a pending PR to disable it for e2e to get more responsive metrics.
https://github.com/GoogleContainerTools/kpt-config-sync/pull/784

Reply:
It's likely the buffer, or batch processing. I'm thinking of using more interactive to decide on ok/error status instead of a hardcoded 1min sleep so that any buffering wouldn't bring flakiness. Might take a few moment.

Reply:
The sleep is replaced by the GCM list request, if request successes within 2min timeout then test move on to check pod log. The buffer is no longer a blocker.